README
NAME
    Test::RDF - Test RDF data for content, validity and equality, etc.

VERSION
    Version 1.19

SYNOPSIS
     use Test::RDF;

     is_valid_rdf($rdf_string, $syntax,  'RDF string is valid according to selected syntax');
     is_rdf($rdf_string, $syntax1, $expected_rdf_string, $syntax2, 'The two strings have the same triples');
     isomorph_graphs($model, $expected_model, 'The two models have the same triples');
     are_subgraphs($model1, $model2, 'Model 1 is a subgraph of model 2' );
     has_uri($uri_string, $model, 'Has correct URI');
     hasnt_uri($uri_string, $model, "Hasn't correct URI");
     has_subject($uri_string, $model, 'Subject URI is found');
     has_predicate($uri_string, $model, 'Predicate URI is found');
     has_object_uri($uri_string, $model, 'Object URI is found');
     has_literal($string, $language, $datatype, $model, 'Literal is found');
     hasnt_literal($string, $language, $datatype, $model, 'Literal is not found');
     pattern_target($model);
     pattern_ok($pattern, '$pattern found in $model');
     pattern_fail($pattern, '$pattern not found in $model');

DESCRIPTION
    This Perl module, Test::RDF, provides tools for testing code which deals
    with RDF. It can test RDF for validity, check if two RDF graphs are the
    same, or subgraphs of each other, if a URI is or is not in a dataset, if
    it has certain subjects, predicates, objects or literals. It can also test
    to see if a full pattern is present or absent.

EXPORT
  is_valid_rdf
    Use to check if the input RDF string is valid in the chosen syntax

  is_rdf
    Use to check if the input RDF strings are isomorphic (i.e. the same).

  isomorph_graphs
    Use to check if the input RDF::Trine::Models have isomorphic graphs.

  are_subgraphs
    Use to check if the first RDF::Trine::Models is a subgraph of the second.

  has_subject
    Check if the string URI passed as first argument is a subject in any of
    the statements given in the model given as second argument.

  has_predicate
    Check if the string URI passed as first argument is a predicate in any of
    the statements given in the model given as second argument.

  has_object_uri
    Check if the string URI passed as first argument is a object in any of the
    statements given in the model given as second argument.

  has_literal
    Check if the string passed as first argument, with corresponding optional
    language and datatype as second and third respectively, is a literal in
    any of the statements given in the model given as fourth argument.

    language and datatype may not occur in the same statement, so the test
    fails if they are both set. If none are used, use `undef`, like e.g.

     has_literal('A test', undef, undef, $model, 'Simple literal');

    A test for a typed literal may be done like

     has_literal('42', undef, 'http://www.w3.org/2001/XMLSchema#integer', $model, 'Just an integer');

    and a language literal like

     has_literal('This is a Another test', 'en', undef, $model, 'Language literal');

  hasnt_literal
    This is like the above, only the opposite: It checks if a literal doesn't
    exist. Like the above, the test will fail if the literal is invalid,
    however.

  has_uri
    Check if the string URI passed as first argument is present in any of the
    statements given in the model given as second argument.

  hasnt_uri
    Check if the string URI passed as first argument is not present in any of
    the statements given in the model given as second argument.

  pattern_target
    Tests that the object passed as its parameter is an RDF::Trine::Model or
    RDF::Trine::Store. That is, tests that it is a valid thing to match basic
    graph patterns against.

    Additionally, this test establishes the target for future `pattern_ok`
    tests.

  pattern_ok
    Tests that the pattern passed matches against the target established by
    `pattern_target`. The pattern may be passed as an RDF::Trine::Pattern, or
    a list of RDF::Trine::Statement objects.

     use Test::RDF;
     use RDF::Trine qw[iri literal blank variable statement];
     use My::Module;

     my $foaf = RDF::Trine::Namespace->new('http://xmlns.com/foaf/0.1/');
     pattern_target(My::Module->get_model); # check isa RDF::Trine::Model
     pattern_ok(
       statement(
         variable('who'),
         $foaf->name,
         literal('Kjetil Kjernsmo')
         ),
       statement(
         variable('who'),
         $foaf->page,
         iri('http://search.cpan.org/~kjetilk/')
         ),
       "Data contains Kjetil's details."
       );

    Note: `pattern_target` must have been tested before any `pattern_ok`
    tests.

  pattern_fail
    The same as above, but tests if the pattern returns no results instead.

NOTE
    Graph isomorphism is a complex problem, so do not attempt to run the
    isomorphism tests on large datasets. For more information see
    <http://en.wikipedia.org/wiki/Graph_isomorphism_problem>.

AUTHOR
    Kjetil Kjernsmo, `<kjetilk at cpan.org>`

BUGS
    Please report any bugs using github
    <https://github.com/kjetilk/Test-RDF/issues>

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Test::RDF

    You may find the Perl and RDF community website <http://www.perlrdf.org/>
    useful.

    You can also look for information at:

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Test-RDF>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Test-RDF>

    *   Search CPAN

        <http://search.cpan.org/dist/Test-RDF/>

    *   MetaCPAN

        <https://metacpan.org/module/Test::RDF>

ACKNOWLEDGEMENTS
    Michael Hendricks wrote the first Test::RDF. The present module is a
    complete rewrite from scratch using Gregory Todd William's
    RDF::Trine::Graph to do the heavy lifting.

    Toby Inkster has submitted the pattern_* functions.

LICENSE AND COPYRIGHT
    Copyright 2010 ABC Startsiden AS. Copyright 2010, 2011, 2012, 2013, 2014
    Kjetil Kjernsmo.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published by
    the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Test-RDF
========

Created:      2010-09-03
Home page:    <https://metacpan.org/module/Test::RDF>
Bug tracker:  <https://github.com/kjetilk/Test-RDF/issues>
Maintainer:   Kjetil Kjernsmo <kjetilk@cpan.org>

1.19	2014-10-15	Cosmetic changes and diagnostics.

 [ Bug Fixes ]
 - Old metadata wasn't overwritten, deleted to allow Dist::Inkt to work.

 [ Other ]
 - Added: Add links to Debian.
 - Removed: Remove inc/

1.18	2014-10-14	Cosmetic changes and diagnostics.

 - Added: Improve diagnostics if a model isn't given.
 - Added: Improve package metadata.
 - Use Dist::Inkt packaging.

1.16	2013-06-27	Bugfix and hasnt_literal.

 [ Bug Fixes ]
 - The new Turtle parser reports errors on a different spot.
 - With the new URI tests, my tests needs to check.

 [ Other ]
 - Added: Add new hasnt_literal test.

1.14	2012-11-21	Scalar::Util strange bug.

 - Updated: Dependency on Scalar::Util is anything but 1.23, set it to
   1.24.

1.11	2012-10-17	Update to RDF::Trine 1.001.

 - Updated: Update test to check the new errors of
   RDF::Trine::Parser::Turtle 1.001.

1.10	2012-09-10	Add pattern_fail.

 - Added: Function to test if pattern doesn't exist.
 - Use RDF-based packaging.

1.00	2012-06-30	Stable release.

 - Added: Add a description to the POD.

0.26	2012-06-11	New hasnt_uri test.

 - Added: Add hasnt uri test.
 - Added: Improve pattern_ok diagnostics.
 - Added: Minor cosmetics.
 - Reorgnize metadata to exploit new modules
 - Use github issue tracker.

0.24	2011-10-10	Pattern tests.

 [ Bug Fixes ]
 - Parser may need an empty sub sometimes.

 [ Other ]
 - Added: Minor cosmetics.
 - Added: New pattern_ok and pattern_target functions.
   Toby Inkster++
 - Added: Test for empty RDF strings

0.22	2011-04-01	Subgraph tests.

 - Added: Add a are_subgraphs function for testing if one graph is a
   subgraph of another.
 - Added: More documentation.
 - Added: Use the RDF::Graph::Trine error function for diagnostics
 - Removed: Remove the now unneeded dependency on Text::Diff::Table.

0.20	2010-12-18	Add functions for testing resources and literals.

 - Added: has_literal function to test if a literal is present.
 - Added: has_object_uri function to test if a URI is present as object.
 - Added: has_predicate function to test if a URI is present as predicate.
 - Added: has_subject function to test if a URI is present as subject.
 - Added: has_uri function to test if a URI is present at all in the graph.
 - Check Redland error messages in addition to Trine.

0.11	2010-11-08	Fix tainting problem.

 - Fix the tainting problem many cpantesters have.
 - Got authorized!.
 - Small cosmetic changes.

0.10	2010-11-06	Complete rewrite.

 - A complete rewrite from scratch using RDF::Trine.
 - Backwards incompatible with previous releases.

0.0.3	2006-13-04	Michael Hendricks old release based on RDF::Redland.


lib/Test/RDF.pm
package Test::RDF;

use warnings;
use strict;

use Carp;
use Test::Builder;
use RDF::Redland;
use base qw( Exporter );

our $VERSION = '0.0.3';
our @EXPORT  = qw( rdf_ok rdf_eq );

my $Test = Test::Builder->new();
RDF::Redland::set_log_handler(\&_log_handler);

sub rdf_ok {
    my ($format, $rdf_source, $message) = @_;
    croak "RDF serialization format required"  if !$format;
    croak "RDF data source required"           if !$rdf_source;

    my $rdf_by_ref = ref($rdf_source) eq 'SCALAR';
    croak "RDF file '$rdf_source' doesn't exist"
        if !$rdf_by_ref && !-e $rdf_source;

    # try to parse the file
    eval {
        my $parser = RDF::Redland::Parser->new($format);
        my $stream;
        if ( $rdf_by_ref ) {
            $stream = $parser->parse_string_as_stream(
                $$rdf_source,
                RDF::Redland::URI->new('http://example.org/'),
            );
        }
        else {
            my $uri = RDF::Redland::URI->new("file:$rdf_source");
            $stream = $parser->parse_as_stream($uri);
        }
        while ( !$stream->end() ) {
            my $statement = $stream->current();
            $stream->next();
        }
    };

    # determine success or failure
    $Test->ok(!$@, $message);
    $Test->diag($@) if $@;
}

sub rdf_eq {

    # create two RDF models in memory
    my @models;
    for my $model_name (qw( a b )) {
        my $storage = RDF::Redland::Storage->new();
        my $model = RDF::Redland::Model->new($storage);
        push @models, $model;

        # validate the format and filename arguments
        my $format     = shift @_;
        my $rdf_source = shift @_;

        croak "RDF serialization format required for part $model_name"
            if !$format;
        croak "RDF data source required for part $model_name"
            if !$rdf_source;

        my $rdf_by_ref = ref($rdf_source) eq 'SCALAR';
        croak "RDF file '$rdf_source' doesn't exist for part $model_name"
            if !$rdf_by_ref && !-e $rdf_source;

        # parse the file contents into the model
        my $parser = RDF::Redland::Parser->new($format);
        if ( $rdf_by_ref ) {
            $parser->parse_string_into_model(
                $$rdf_source,
                RDF::Redland::URI->new('http://example.org/'),
                $model,
            );
        }
        else {
            my $uri = RDF::Redland::URI->new("file:$rdf_source");
            $parser->parse_into_model( $uri, $uri, $model );
        }
    }

    my $message = shift @_;

    # are the models the same size?
    if ( $models[0]->size() != $models[1]->size() ) {
        $Test->ok(0, $message);
        $Test->diag(
            'Graphs differ in size: '
            . $models[0]->size() . ' vs. ' . $models[1]->size()
        );
        return;
    }

    # start the comparison with blank bnode maps
    _clear_bnode_mappings();

    # check all the statements for equivalence in the second model
    my $stream = $models[0]->as_stream();
    STATEMENT:
    while( !$stream->end() ) {
        my $stmt = $stream->current();
        next STATEMENT if _has_equivalent_statement($models[1], $stmt);
        $Test->ok(0, $message);
        $Test->diag(
            'Model B is missing the statement ' . $stmt->as_string()
        );
        return;
    }
    continue {
        $stream->next();
    }

    $Test->ok(1, $message);
}

sub _has_equivalent_statement {
    my ($model, $stmt) = @_;

    my $subject   = $stmt->subject();
    my $predicate = $stmt->predicate();
    my $object    = $stmt->object();
    my ($subject_is_floating, $object_is_floating) = (0)x2;

    # anchor the subject if possible
    if ( $subject->is_blank() ) {
        $subject_is_floating = 1;
        if ( my $mapped = _get_bnode_mapping($subject) ) {
            $subject = $mapped;
            $subject_is_floating = 0;
        }
    }

    # anchor the object if possible
    if ( $object->is_blank() ) {
        $object_is_floating = 1;
        if ( my $mapped = _get_bnode_mapping($object) ) {
            $object = $mapped;
            $object_is_floating = 0;
        }
    }

    if ( $subject_is_floating && $object_is_floating ) {
        my @candidates = $model->find_statements(
            RDF::Redland::Statement->new( undef, $predicate, undef )
        );
        for my $candidate (@candidates) {
            next if !$candidate->subject->is_blank();
            next if !$candidate->object->is_blank();
            # TODO what if $subject and $object are the same bnode?
            next if !_is_valid_mapping( $subject, $candidate->subject() );
            next if !_is_valid_mapping( $object, $candidate->object() );
            _create_bnode_mapping( $subject, $candidate->subject() );
            _create_bnode_mapping( $object, $candidate->object() );
            $model->remove_statement($candidate);
            return 1;
        }
    }
    elsif ( $subject_is_floating ) {
        my @sources = $model->sources( $predicate, $object );
        for my $source (@sources) {
            next if !$source->is_blank();
            next if !_is_valid_mapping( $subject, $source );
            _create_bnode_mapping( $subject, $source );
            $model->remove_statement( $source, $predicate, $object );
            return 1;
        }
    }
    elsif ( $object_is_floating ) {
        my @targets = $model->targets( $subject, $predicate );
        for my $target (@targets) {
            next if !$target->is_blank();
            next if !_is_valid_mapping( $object, $target );
            _create_bnode_mapping( $object, $target );
            $model->remove_statement( $subject, $predicate, $target );
            return 1;
        }
    }
    else {
        my $anchored_stmt
            = RDF::Redland::Statement->new( $subject, $predicate, $object );
        if ( $model->contains_statement($anchored_stmt) ) {
            $model->remove_statement($anchored_stmt);
            return 1;
        }
    }

    return;  # no equivalent statement found
}

{
    # map bnodes between models to insure that we get
    # a true bijection when calculating rdf_eq()
    my %a_to_b;
    my %b_to_a;
    sub _clear_bnode_mappings {
        %a_to_b = ();
        %b_to_a = ();
    }
    sub _get_bnode_mapping {
        my ($bnode_a) = @_;
        return $a_to_b{ $bnode_a->blank_identifier() };
    }
    sub _is_valid_mapping {
        my ( $bnode_a, $bnode_b ) = @_;
        return if $a_to_b{ $bnode_a->blank_identifier() };
        return if $b_to_a{ $bnode_b->blank_identifier() };
        return 1;
    }
    sub _create_bnode_mapping {
        my ( $bnode_a, $bnode_b ) = @_;
        die "Invalid mapping" if !_is_valid_mapping( $bnode_a, $bnode_b );
        $a_to_b{ $bnode_a->blank_identifier() } = $bnode_b;
        $b_to_a{ $bnode_b->blank_identifier() } = $bnode_a;
    }
}

# Handle error and warning messages from Redland, cleaning
# up the messages as necessary
sub _log_handler {
    my (
        $code,   $level, $facility, $message, $line,
        $column, $byte,  $file,     $uri
    ) = @_;

    die "$message\n";
}

1;

__END__

=head1 NAME
 
Test::RDF - Test RDF data for validity and equality
 
 
=head1 VERSION
 
This documentation refers to Test::RDF version 0.0.3
 
 
=head1 SYNOPSIS
 
    use Test::More tests => 3;
    use Test::RDF;
    
    rdf_ok( rdfxml => 'data.rdf', 'data validity' );
    rdf_eq( rdfxml => 'data.rdf', turtle => 'data.ttl', 'XML==Turtle' );
    
    rdf_eq(
        ntriples => \'_:a <http://example.org> "literal .',
        turtle   => \' [] <http://example.org> "literal .',
        'ntriples and turtle blank node equivalence',
    );

=head1 DESCRIPTION

Test::RDF is used for testing RDF data in various formats.  Currently,
Test::RDF exports two functions L</rdf_ok> (check the validity of various RDF
serialization formats) and L</rdf_eq> (check for RDF graph equivalence).

=head1 SUBROUTINES

=head2 rdf_eq

    Arguments: $FORMAT, $SOURCE, $FORMAT, $SOURCE [, $MESSAGE]

Compares the RDF graphs created by the two RDF serializations for graph
equivalence.  RDF graph equivalence is defined by the RDF Concepts and
Abstract Syntax document here:
L<http://www.w3.org/TR/rdf-concepts/#section-graph-equality>.  If the two
graphs are equivalent, the test passes.  If the two graphs are not equivalent,
the test fails with a helpful diagnostic message.

The C<$FORMAT> arguments should be one of: C<rdfxml>, C<turtle> or C<ntriples>
(actually, you can use any format allowed by your version of
L<RDF::Redland::Parser>).  C<$SOURCE> should be either the path to a file or a
reference to a scalar containing RDF data in the specified format.
C<$MESSAGE> is an optional message to use when displaying the "ok" or "not ok"
message.

C<rdf_eq> does not correctly handle reflexive statements involving bnodes.
That is, statements where subject and object are the same blank node.
 
=head2 rdf_ok

    Arguments: $FORMAT, $SOURCE [, $MESSAGE]

C<$FORMAT> specifies the expected format of the RDF file.  It should be one
of: C<rdfxml>, C<turtle> or C<ntriples> (actually, you can use any format
allowed by your version of L<RDF::Redland::Parser> but those three are the
most commonly useful).  C<$SOURCE> should be either the path to a file or a
reference to a scalar containing RDF data in the specified format.
C<$MESSAGE> is an optional message to use when displaying the "ok" or "not ok"
message.
 
=head1 CONFIGURATION AND ENVIRONMENT
 
Test::RDF requires no configuration files or environment variables.
 
=head1 DEPENDENCIES
 
=over

=item *

RDF::Redland

=back
 
=head1 INCOMPATIBILITIES
 
None known
 
=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests to
C<bug-test-rdf at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Test-RDF>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Test::RDF

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Test-RDF>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Test-RDF>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Test-RDF>

=item * Search CPAN

L<http://search.cpan.org/dist/Test-RDF>

=back

=head1 ACKNOWLEDGEMENTS

Dave Beckett for Redland.

=head1 AUTHOR

Michael Hendricks  <michael@palmcluster.org>

=head1 LICENSE AND COPYRIGHT
 
Copyright (c) 2006 Michael Hendricks (<michael@palmcluster.org>). All rights
reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.
 


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "Test RDF data for validity and equality.",
  "author"         => ["Kjetil Kjernsmo <kjetilk\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.023, CPAN::Meta::Converter version 2.142060",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Test-RDF",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        runtime => {
                          requires => {
                            "RDF::Trine"            => 0.134,
                            "Scalar::Util"          => 1.24,
                            "Test::Builder::Module" => 0,
                            "Test::More"            => 0,
                          },
                        },
                        test => { requires => { "Test::Tester" => 0 } },
                      },
  "provides"       => { "Test::RDF" => { file => "lib/Test/RDF.pm", version => 1.19 } },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => { web => "https://github.com/kjetilk/Test-RDF/issues" },
                        homepage     => "https://metacpan.org/module/Test::RDF",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => { type => "git", web => "https://github.com/kjetilk/Test-RDF/" },
                        x_identifier => "http://purl.org/NET/cpan-uri/dist/Test-RDF/project",
                        x_IRC        => "irc://irc.perl.org/#perlrdf",
                      },
  "version"        => 1.19,
  "x_contributors" => ["Toby Inkster"],
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

